client/session.go

```diff
@@ -548,21 +556,90 @@ func (s *session) newHostQueue(host topology.Host, topologyMap topology.Map) hos
 	return hostQueue
 }
 
+type rc struct {
+	d func()
+	n int32
+}
+
+func (r *rc) incref() {
+	if atomic.AddInt32(&r.n, 1) <= 0 {
+		panic("invalid rc state")
+	}
+}
+
+func (r *rc) decref() {
+	if v := atomic.AddInt32(&r.n, -1); v == 0 {
+		r.d()
+	} else if v < 0 {
+		panic("invalid rc state")
+	}
+}
+
+type writeState struct {
```

Comment:
Can we move `writeState` and `rc` subtypes into another file?  I'll put up a PR soon to start pulling other subtypes from `session.go ` into their own files.  This file is getting too monolithic.